### PR TITLE
Add Lisht Activation Function in reference to issue #2181.

### PR DESCRIPTION
**Pull request by kartikdutt18**
Hi everyone, I have added implementation of LiSHT function in mlpack. LiSHT function already exited in tensor flow and it was used as source for testing.
The testing script can be found [here](https://colab.research.google.com/drive/10nrsslrKaz-wmjyWMAHLL-MmIwFUX-Yy).
The reference paper can be found [here](https://arxiv.org/pdf/1901.05894.pdf).
To quote from the paper,

> It is observed that the proposed LiSHT activation function outperforms the other activation functions with a record ac- curacies of 99.59% and 92.92%, and 75.32% over MNIST, CIFAR-10 and CIFAR-100 datasets, respectively.

> The unbounded, symmetric and more non- linear properties of the proposed LiSHT activation function facilitates better and efficient training as compared to the other activation functions such as Tanh, ReLU and Swish.

> The sentiment classification performance in terms of the validation accuracy is reported in Table III over twitter140 dataset with LSTM for different activations. It is observed that the performance of proposed LiSHT activation function is better than Tanh and Swish, whereas the same as ReLU. It points out one important observation that by considering the negative values as negative by Swish degrades the per- formance because it leads the Swish activation more towards the linear function as compared to the ReLU activation

### CODE REVIEW

File: src/mlpack/methods/ann/activation_functions/lisht_function.hpp [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+/**
+ * The Lisht function, defined by
+ *
+ * @f{eqnarray*}{
+ * f(x) = x * tanh(x)
+ * f'(x) = tanh(x) + x * (1 - tanh^{2}(x))
+ * @f}
+ */
+class LishtFunction
```
Comment (zoq) [inline src/mlpack/methods/ann/activation_functions/lisht_function.hpp]: Perhaps we should use the same spelling as used in the paper: LiSHT?

File: src/mlpack/methods/ann/activation_functions/lisht_function.hpp [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+
+  /**
+   * Computes the first derivative of the Lisht function.
+   *
+   * @param y Input data.
+   * @return f'(x)
+   */
+  static double Deriv(const double y)
+  {
+    return (4 * y * std::exp(2 * y) + std::exp(4 * y) - 1) /
```
Comment (zoq) [inline src/mlpack/methods/ann/activation_functions/lisht_function.hpp]: Is this some kind of approximation for the derivative as described in the paper?

Comment (kartikdutt18) [reply src/mlpack/methods/ann/activation_functions/lisht_function.hpp]: Hi, no this is the same expression as that in the paper but simplified further in exponentials to make it a bit faster.

File: src/mlpack/methods/ann/activation_functions/lisht_function.hpp (hunk 1 above)
Comment (kartikdutt18) [reply src/mlpack/methods/ann/activation_functions/lisht_function.hpp]: Hi @zoq, I will make the changes.

**Comment by kartikdutt18** [comment]
Hi @zoq, I have made the changes to rename it to LiSHT. Have a look whenever you get the chance. Thanks.

**Comment by kartikdutt18** [comment]
Hi @zoq, in the test case link in description I have used the paper's definition to verify that my simplification is correct. Thanks.

File: src/mlpack/methods/ann/activation_functions/lisht_function.hpp (hunk 2 above)
Comment (zoq) [reply src/mlpack/methods/ann/activation_functions/lisht_function.hpp]: Thanks for the clarification.

Comment (zoq) [reply src/mlpack/methods/ann/activation_functions/lisht_function.hpp]: Do you think this is faster as using `tanh`? I mean if I would use `tanh` the expression would look at lot simpler, but if that one is faster, I don't mind to use that instead but in this case let's reference the formula from the paper right above.

Comment (kartikdutt18) [reply src/mlpack/methods/ann/activation_functions/lisht_function.hpp]: Hi @zoq, I am changing to paper's definition here since I ran a test for both to compare time to get following results on an input of 100000000
Time: 8.66564
Time: 8.68422
Which I think is within in margin of error and the definition in paper looks cleaner. I thought it must be much faster as it was [here.](https://github.com/digantamisra98/Mish/issues/22#issuecomment-573394994)

Comment (zoq) [reply src/mlpack/methods/ann/activation_functions/lisht_function.hpp]: Agreed, the current implementation looks cleaner, I think the compiler already does a pretty good job at optimizing `tanh`.

**Comment by kartikdutt18** [comment]
Thanks @zoq, for fixing the merge conflict.

**Comment by zoq** [comment]
Thanks for another great contribution.

**Comment by kartikdutt18** [comment]
Thanks @zoq, for all the help and reviews.
